pkg/maps/lbmap/lbmap.go

```diff
@@ -425,152 +387,85 @@ func UpdateService(fe ServiceKey, backends []ServiceValue, addRevNAT bool, revNA
-		logfields.Object:    logfields.Repr(svcValue),
-	}).Debug("converting ServiceKey and ServiceValue to frontend and backend")
-
-	if svcKey.IsIPv6() {
-		svc6Val := svcValue.(*Service6Value)
-		svcID = loadbalancer.ServiceID(svc6Val.RevNat)
-		beIP = svc6Val.Address.IP()
-		bePort = svc6Val.Port
-		beWeight = svc6Val.Weight
-	} else {
-		svc4Val := svcValue.(*Service4Value)
-		svcID = loadbalancer.ServiceID(svc4Val.RevNat)
-		beIP = svc4Val.Address.IP()
-		bePort = svc4Val.Port
-		beWeight = svc4Val.Weight
+	err = updateMasterServiceV2(svcKeyV2, len(svc.backendsV2), nNonZeroWeights, revNATID)
+	if err != nil {
+		return fmt.Errorf("unable to update service %+v: %s", svcKeyV2, err)
 	}
 
-	feL3n4Addr := serviceKey2L3n4Addr(svcKey)
-	beLBBackEnd := loadbalancer.NewLBBackEnd(loadbalancer.NONE, beIP, bePort, beWeight)
-
-	feL3n4AddrID := &loadbalancer.L3n4AddrID{
-		L3n4Addr: *feL3n4Addr,
-		ID:       svcID,
+	err = updateWrrSeqV2(svcKeyV2, weights)
+	if err != nil {
+		return fmt.Errorf("unable to update service weights for %s with value %+v: %s", svcKeyV2.String(), weights, err)
+	}
+	for i := slot; i <= existingCount; i++ {
+		svcKeyV2.SetSlave(i)
+		if err := deleteServiceLockedV2(svcKeyV2); err != nil {
+			return fmt.Errorf("unable to delete service %+v: %s", svcKeyV2, err)
+		}
+		log.WithFields(logrus.Fields{
+			logfields.SlaveSlot:  i,
+			logfields.ServiceKey: svcKeyV2,
+		}).Debug("Deleted service entry")
 	}
 
-	return feL3n4AddrID, beLBBackEnd
-}
-
-// RevNat6Value2L3n4Addr converts the given RevNat6Value to a L3n4Addr.
-func revNat6Value2L3n4Addr(revNATV *RevNat6Value) *loadbalancer.L3n4Addr {
-	return loadbalancer.NewL3n4Addr(loadbalancer.NONE, revNATV.Address.IP(), revNATV.Port)
-}
-
-// revNat4Value2L3n4Addr converts the given RevNat4Value to a L3n4Addr.
-func revNat4Value2L3n4Addr(revNATV *RevNat4Value) *loadbalancer.L3n4Addr {
-	return loadbalancer.NewL3n4Addr(loadbalancer.NONE, revNATV.Address.IP(), revNATV.Port)
-}
-
-// revNatValue2L3n4AddrID converts the given RevNatKey and RevNatValue to a L3n4AddrID.
-func revNatValue2L3n4AddrID(revNATKey RevNatKey, revNATValue RevNatValue) *loadbalancer.L3n4AddrID {
-	var (
-		svcID loadbalancer.ServiceID
-		be    *loadbalancer.L3n4Addr
-	)
-	if revNATKey.IsIPv6() {
-		revNat6Key := revNATKey.(*RevNat6Key)
-		svcID = loadbalancer.ServiceID(revNat6Key.Key)
-
-		revNat6Value := revNATValue.(*RevNat6Value)
-		be = revNat6Value2L3n4Addr(revNat6Value)
+	// Delete no longer needed backends
+	if fe.IsIPv6() {
+		backendKey = NewBackend6Key(0)
 	} else {
-		revNat4Key := revNATKey.(*RevNat4Key)
-		svcID = loadbalancer.ServiceID(revNat4Key.Key)
+		backendKey = NewBackend4Key(0)
+	}
 
-		revNat4Value := revNATValue.(*RevNat4Value)
-		be = revNat4Value2L3n4Addr(revNat4Value)
+	for _, backendID := range removedBackendIDs {
+		backendKey.SetID(backendID)
+		if err := deleteBackendLocked(backendKey); err != nil {
+			return fmt.Errorf("Unable to delete backend with ID %d: %s", backendID, err)
```

Comment:
Should this continue attempting to remove other backends even if one fails?

Reply:
Yep, but it's probably part of https://github.com/cilium/cilium/pull/7540#discussion_r271785436.